server/routes/prediction.routes.js

```diff
@@ -359,146 +364,155 @@ function normalizeMatchFilter(filter, field){
  * @return {Promise<Array(Prediction)>} All predictions that match the filter
  */
 /** @namespace filter.numDocs */
-async function getPredictions (filter, user) {
 
-  try {
-    let first = filter.first || 0
-    let max_fetch_rows = filter.rows || configuration.getConfig("defaultMaxPredictions", 1000)
 
+async function getPredictions(filter, user) {
+  try {
+    logger.debug('GetPredictions - Start', { filter, user });
+    let first = filter.first || 0;
+    let max_fetch_rows = filter.rows || configuration.getConfig('defaultMaxPredictions', 1000);
 
-    if ( user === undefined || user.agency === undefined || user.userRole === undefined ) {
-      return []
+    // Ensure user has necessary information
+    if (!user?.agency || !user?.userRole || !user?.email) {
+      logger.warn('Missing user info');
+      return { predictions: [], first: 0, rows: 0, totalCount: 0 };
     }
 
-    logger.debug("Entering getPredictions")
-    // await updatePredictionTable()
+    // Compute the user's office dynamically from their email
+    user.office = getOfficeFromEmail(user.email);
+    logger.debug('Computed user office', { userOffice: user.office });
 
     let attributes = {
       offset: first,
       limit: max_fetch_rows,
-      include: [{
-        model: SurveyResponse,
-        as: 'feedback'
-      }],
-    }
+      include: [{ model: SurveyResponse, as: 'feedback' }],
+      where: {},
+    };
+    logger.debug('Initial attributes', { attributes });
 
-    // filter to allowed notice types
-    let types = configuration.getConfig("VisibleNoticeTypes", ['Solicitation', 'Combined Synopsis/Solicitation', 'RFQ'])
-    
-    attributes.where = {
-      noticeType: {
-        [Op.in]: types
-      }
+    // Notice type filter
+    let types = configuration.getConfig('VisibleNoticeTypes', ['Solicitation', 'Combined Synopsis/Solicitation', 'RFQ']);
+    if (types?.length) {
+      attributes.where.noticeType = types; // Simplified without Op.in
     }
 
-    // filter out rows
+    // Global text search
     if (filter.globalFilter) {
-      attributes.where.searchText = { [Op.like]: `%${filter.globalFilter.toLowerCase()}%` }
-    }
-    for (let f of ['office', 'agency', 'title', 'solNum', 'reviewRec', 'id']) {
-      normalizeMatchFilter(filter, f)
+      attributes.where.searchText = { [Sequelize.Op.like]: `%${filter.globalFilter.toLowerCase()}%` };
     }
 
-
-    // process PrimeNG filters: filter.filters = { field: { value: 'x', matchMode: 'equals' } }
-    if (filter.filters) {
-      for (let f in filter.filters) {
-        if (filter.filters.hasOwnProperty(f) && filter.filters[f].matchMode === 'equals') {
-          attributes.where[f] = {[Op.eq]: filter.filters[f].value}
-        }
+    // Normalize filters
+    ['office', 'agency', 'title', 'solNum', 'reviewRec', 'id'].forEach(f => {
```

Comment:
I personally prefer separating out code into the small functions instead of having a large function with a lot of executing code.

Reply:
Thank you for this comment, I fell back to how I used to do things by making a giant, messy function.

The code has been completely refactored into smaller, focused functions:
* `applyBasicFilters` for notice types and global search
* `normalizeStandardFilters` for filter normalization
* `applyRegularFilters` for standard field filters
* `applyDateFilters` for date filtering
* `applyAgencyOfficeFilters` for agency/office logic
* `applySorting` for sort ordering

Each function has a single responsibility, making the code more maintainable and testable.